rust/lance-index/src/vector/v3/storage.rs

```diff
@@ -100,3 +114,113 @@ impl<Q: Quantization> StorageBuilder<Q> {
+            + self.metadata.deep_size_of_children(context)
+            + self.ivf.deep_size_of_children(context)
+    }
+}
+
+// impl<Q: Quantization> Clone for IvfQuantizationStorage<Q> {
+//     fn clone(&self) -> Self {
+//         Self {
+//             reader: self.reader.clone(),
+//             metric_type: self.metric_type,
+//             quantizer: self.quantizer.clone(),
+//             metadata: self.metadata.clone(),
+//             ivf: self.ivf.clone(),
+//         }
+//     }
+// }
+
+#[allow(dead_code)]
+impl<Q: Quantization> IvfQuantizationStorage<Q> {
+    /// Open a Loader.
+    ///
+    ///
+    pub async fn open(reader: FileReader) -> Result<Self> {
+        let schema = reader.schema();
+
+        let distance_type = DistanceType::try_from(
+            schema
+                .metadata
+                .get(DISTANCE_TYPE_KEY)
+                .ok_or(Error::Index {
+                    message: format!("{} not found", INDEX_METADATA_SCHEMA_KEY),
+                    location: location!(),
+                })?
+                .as_str(),
+        )?;
+
+        let ivf_pb_bytes =
+            hex::decode(schema.metadata.get(IVF_METADATA_KEY).ok_or(Error::Index {
+                message: format!("{} not found", IVF_METADATA_KEY),
+                location: location!(),
+            })?)
+            .map_err(|e| Error::Index {
+                message: format!("Failed to decode IVF metadata: {}", e),
+                location: location!(),
+            })?;
+        let ivf = IvfData::try_from(pb::Ivf::decode(ivf_pb_bytes.as_ref())?)?;
+
+        let quantizer_metadata: Q::Metadata = serde_json::from_str(
+            schema
+                .metadata
+                .get(Q::metadata_key())
+                .ok_or(Error::Index {
+                    message: format!("{} not found", Q::metadata_key()),
+                    location: location!(),
+                })?
+                .as_str(),
+        )?;
+        let quantizer = Q::from_metadata(&quantizer_metadata, distance_type)?;
+        Ok(Self {
+            reader,
+            distance_type,
+            quantizer,
+            metadata: quantizer_metadata,
+            ivf,
+        })
+    }
+
+    pub fn quantizer(&self) -> &Quantizer {
+        &self.quantizer
+    }
+
+    pub fn metadata(&self) -> &Q::Metadata {
+        &self.metadata
+    }
+
+    /// Get the number of partitions in the storage.
+    pub fn num_partitions(&self) -> usize {
+        self.ivf.num_partitions()
+    }
+
+    pub async fn load_partition(&self, part_id: usize) -> Result<Q::Storage> {
+        let range = self.ivf.row_range(part_id);
+        let batches = self
+            .reader
+            .read_stream(ReadBatchParams::Range(range), 4096, 16)?
+            .peekable()
+            .try_collect::<Vec<_>>()
+            .await?;
+        let schema = Arc::new(self.reader.schema().as_ref().into());
+        let batch = concat_batches(&schema, batches.iter())?;
```

Comment:
If the goal is to load one giant batch then use a larger `batch_size`?  You should be able to use at least `u32::MAX`.

Reply:
This